pkg/hubble/exporter/exporter.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exporter
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	observerAPI "github.com/cilium/cilium/api/v1/observer"
+	"github.com/cilium/cilium/pkg/hubble/exporter/exporteroption"
+
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc/metadata"
+	"gopkg.in/natefinch/lumberjack.v2"
+)
+
+var _ observerAPI.Observer_GetFlowsServer = (*exporter)(nil)
+
+// exporter implements Observer_GetFlowsServer interface. Instead of sending responses over
+// a socket it writes them to a file in JSON format.
+type exporter struct {
+	logger   logrus.FieldLogger
+	observer observerAPI.ObserverServer
+	encoder  *json.Encoder
+}
+
+// NewExporter initializes an exporter.
+func NewExporter(
+	logger logrus.FieldLogger,
+	observer observerAPI.ObserverServer,
+	options ...exporteroption.Option) (*exporter, error) {
+	opts := exporteroption.Default // start with defaults
+	for _, opt := range options {
+		if err := opt(&opts); err != nil {
+			return nil, fmt.Errorf("failed to apply option: %v", err)
+		}
+	}
+	logger.WithField("options", opts).Info("Configuring Hubble event exporter")
+	encoder := json.NewEncoder(&lumberjack.Logger{
+		Filename:   opts.Path,
+		MaxSize:    opts.MaxSizeMB,
+		MaxBackups: opts.MaxBackups,
+		Compress:   opts.Compress,
+	})
+	return newExporter(logger, observer, encoder), nil
+}
+
+func newExporter(logger logrus.FieldLogger, observer observerAPI.ObserverServer, encoder *json.Encoder) *exporter {
+	return &exporter{
+		logger:   logger,
+		observer: observer,
+		encoder:  encoder,
+	}
+}
+
+// Start calls GetFlows and writes responses to a file.
+func (e *exporter) Start() {
+	// Wait for Hubble to observe at least 2 flows to prevent ring.LastWriteParallel()
+	// from underflowing.
+	for {
+		status, err := e.observer.ServerStatus(context.Background(), &observerAPI.ServerStatusRequest{})
+		if err == nil && status.NumFlows > 1 {
+			break
+		}
+		e.logger.WithError(err).Debug("Waiting for Hubble to start...")
+		time.Sleep(1 * time.Second)
+	}
+	req := observerAPI.GetFlowsRequest{Follow: true}
```

Comment:
This seems to be something that's rather high overhead; we're encoding/decoding proto, then encoding it into json to write to disk. 

Doesn't it make sense to make this whole functionality a `OnDecodedFlow` server option?

Reply:
yeah i thought about that,`OnDecodedFlow` only gets called for flows, and not for other events like agent_event. we could add another hook here if you prefer: https://github.com/michi-covalent/cilium/blob/28309a9cfd4cf38dceb86ca58fc20b5e460d112f/pkg/hubble/observer/local_observer.go#L164 that way at least we are bypassing reading events from ring buffer.

but either way events are already decoded here: https://github.com/michi-covalent/cilium/blob/28309a9cfd4cf38dceb86ca58fc20b5e460d112f/pkg/hubble/observer/local_observer.go#L138

undecoded events don't have all the extra metadata like labels so they are not as useful imo.

Reply:
added `OnDecodedEvent` hook.